Segfault when using non existent image

I stumbled over this, when I tried the code example and supplied a filename which didn't exist:
```go
func main() {
	if len(os.Args) < 3 {
		log.Fatalln("Usage: hash img1 img2")
	}

	log.Printf("1: %s, 2: %s\n", os.Args[1], os.Args[2])
	file1, _ := os.Open(os.Args[1])
	file2, _ := os.Open(os.Args[2])
	defer file1.Close()
	defer file2.Close()

	img1, _ := jpeg.Decode(file1)
	img2, _ := jpeg.Decode(file2)

	hash1, _ := goimagehash.DifferenceHash(img1)
	hash2, _ := goimagehash.DifferenceHash(img2)
	distance, _ := hash1.Distance(hash2)

	fmt.Printf("Distance between images: %v\n", distance)
}
```
Whenn I call this with the first or second argument given a non-existent file, the following happens:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x4ab74c]

goroutine 1 [running]:
github.com/corona10/goimagehash.(*ImageHash).GetKind(...)
        /home/scip/go/pkg/mod/github.com/corona10/goimagehash@v1.1.0/imagehash.go:75
github.com/corona10/goimagehash.(*ImageHash).Distance(...)
        /home/scip/go/pkg/mod/github.com/corona10/goimagehash@v1.1.0/imagehash.go:57
main.main()
        /home/scip/dev/play/hash/main.go:28 +0x2cc
```

As you can see, there's no error checking on my side, so that's not the librarys fault. However, I think it would be better to do some additional check. There's already a nil test in DifferenceHash(), so I don't know what's going on exactly, but a viable error message would be better.

Thanks for the library anyway!

Tom

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -4,7 +4,7 @@ jobs:
   test:
     strategy:
       matrix:
-        go-version: [1.6.x, 1.7.x, 1.8.x, 1.9.x, 1.10.x, 1.11.x, 1.12.x, 1.13.x, 1.14.x, 1.15.x, 1.16.x, 1.17.x, 1.18.x, 1.19.x, 1.20.x]
+        go-version: [1.6.x, 1.7.x, 1.8.x, 1.9.x, 1.10.x, 1.11.x, 1.12.x, 1.13.x, 1.14.x, 1.15.x, 1.16.x, 1.17.x, 1.18.x, 1.19.x, 1.20.x, 1.21.x]
         platform: [ubuntu-latest]
     runs-on: ${{ matrix.platform }}
     steps:

--- a/.github/workflows/ci_gomodule.yml
+++ b/.github/workflows/ci_gomodule.yml
@@ -4,7 +4,7 @@ jobs:
   test:
     strategy:
       matrix:
-        go-version: [1.11.x, 1.12.x, 1.13.x, 1.14.x, 1.15.x, 1.16.x, 1.17.x, 1.18.x, 1.19.x, 1.20.x]
+        go-version: [1.11.x, 1.12.x, 1.13.x, 1.14.x, 1.15.x, 1.16.x, 1.17.x, 1.18.x, 1.19.x, 1.20.x, 1.21.x]
         platform: [ubuntu-latest, macos-latest, windows-latest]
     runs-on: ${{ matrix.platform }}
     steps:

--- a/imagehash.go
+++ b/imagehash.go
@@ -13,6 +13,8 @@ import (
 	"io"
 )
 
+var errNoOther = errors.New("other should not be nil")
+
 // Kind describes the kinds of hash.
 type Kind int
 
@@ -55,7 +57,7 @@ func (h *ImageHash) Bits() int {
 // Distance method returns a distance between two hashes.
 func (h *ImageHash) Distance(other *ImageHash) (int, error) {
 	if other == nil {
-		return -1, errors.New("other should not be nil")
+		return -1, errNoOther
 	}
 	if h.GetKind() != other.GetKind() {
 		return -1, errors.New("Image hashes's kind should be identical")

--- a/imagehash.go
+++ b/imagehash.go
@@ -54,6 +54,9 @@ func (h *ImageHash) Bits() int {
 
 // Distance method returns a distance between two hashes.
 func (h *ImageHash) Distance(other *ImageHash) (int, error) {
+	if other == nil {
+		return -1, errors.New("other should not be nil")
+	}
 	if h.GetKind() != other.GetKind() {
 		return -1, errors.New("Image hashes's kind should be identical")
 	}

--- a/imagehash_test.go
+++ b/imagehash_test.go
@@ -56,6 +56,18 @@ func TestNewImageHash(t *testing.T) {
 	}
 }
 
+func TestNil(t *testing.T) {
+	hash := NewImageHash(0, AHash)
+	dis, err := hash.Distance(nil)
+	if dis != -1 {
+		t.Errorf("Distance is expected as %d but got %d", -1, dis)
+	}
+	expectedError := errors.New("other should not be nil")
+	if err != nil && err.Error() != expectedError.Error() {
+		t.Errorf("Expected err %s, actual %s", expectedError, err)
+	}
+}
+
 func TestSerialization(t *testing.T) {
 	checkErr := func(err error) {
 		if err != nil {

--- a/imagehash_test.go
+++ b/imagehash_test.go
@@ -59,12 +59,12 @@ func TestNewImageHash(t *testing.T) {
 func TestNil(t *testing.T) {
 	hash := NewImageHash(0, AHash)
 	dis, err := hash.Distance(nil)
-	if dis != -1 {
-		t.Errorf("Distance is expected as %d but got %d", -1, dis)
-	}
 	expectedError := errors.New("other should not be nil")
 	if err != nil && err.Error() != expectedError.Error() {
 		t.Errorf("Expected err %s, actual %s", expectedError, err)
+	}
+	if dis != -1 {
+		t.Errorf("Distance is expected as %d but got %d", -1, dis)
 	}
 }
 

--- a/imagehash_test.go
+++ b/imagehash_test.go
@@ -59,9 +59,8 @@ func TestNewImageHash(t *testing.T) {
 func TestNil(t *testing.T) {
 	hash := NewImageHash(0, AHash)
 	dis, err := hash.Distance(nil)
-	expectedError := errors.New("other should not be nil")
-	if err != nil && err.Error() != expectedError.Error() {
-		t.Errorf("Expected err %s, actual %s", expectedError, err)
+	if !errors.Is(err, errNoOther) {
+		t.Errorf("Expected err %s, actual %s", errNoOther, err)
 	}
 	if dis != -1 {
 		t.Errorf("Distance is expected as %d but got %d", -1, dis)

--- a/imagehash_test.go
+++ b/imagehash_test.go
@@ -59,7 +59,7 @@ func TestNewImageHash(t *testing.T) {
 func TestNil(t *testing.T) {
 	hash := NewImageHash(0, AHash)
 	dis, err := hash.Distance(nil)
-	if !errors.Is(err, errNoOther) {
+	if err != errNoOther {
 		t.Errorf("Expected err %s, actual %s", errNoOther, err)
 	}
 	if dis != -1 {

